### ISSUE TITLE
OSMC remote has issue with context menu on Kodi 18

### ISSUE THREAD

**Issue by dagwieers** [osmc/osmc#478]
I know this is likely the wrong forum to report this, but it may be relevant to OSMC itself too.

In LibreELEC 8.90.X using Kodi 18 the OSMC remote context key (⋮☰) no longer maps to the context menu, but now maps to TV Timeline. Which is quite frustrating as the context key is an important one.

Is this a known issue i.e. with OSMC ? Is this specific to LibreELEC or a more general Kodi 18 issue/regression ?

**Comment by dagwieers** [comment]
After experimenting with my system, I figured out that by stopping the **eventlircd** service, everything worked again as expected.

So I only needed to run:
```
sudo systemctl stop eventlircd
sudo systemctl disable eventlircd
```

**Comment by dagwieers** [comment]
From the eventlircd configuration I do find:
```
kodi01:/ # cat ./etc/eventlircd.d/osmc_rf2.evmap
#KEY_HOME		= KEY_HOME
KEY_I			= KEY_INFO
#KEY_UP			= KEY_UP
#KEY_DOWN		= KEY_DOWN
#KEY_LEFT		= KEY_LEFT
#KEY_RIGHT		= KEY_RIGHT
KEY_ENTER		= KEY_OK
#KEY_BACK		= KEY_BACK
KEY_C			= KEY_EPG
#KEY_PLAYPAUSE		= KEY_PLAYPAUSE
#KEY_STOP		= KEY_STOP
KEY_MINUS		= KEY_VOLUMEDOWN
KEY_EQUAL		= KEY_VOLUMEUP
kodi01:/ # cat ./etc/eventlircd.d/osmc_rf.evmap
#KEY_HOME		= KEY_HOME
#KEY_INFO		= KEY_INFO
#KEY_UP			= KEY_UP
#KEY_DOWN		= KEY_DOWN
#KEY_LEFT		= KEY_LEFT
#KEY_RIGHT		= KEY_RIGHT
KEY_ENTER		= KEY_OK
#KEY_BACK		= KEY_BACK
KEY_CONTEXT_MENU	= KEY_EPG
#KEY_PLAYPAUSE		= KEY_PLAYPAUSE
#KEY_STOP		= KEY_STOP
#KEY_REWIND		= KEY_REWIND
#KEY_FASTFORWARD	= KEY_FASTFORWARD
```
Which does map the Context key to the EPG, so something must have changed with the eventlircd shipping with LibreELEC 8.90 ?

**Comment by samnazarko** [comment]
This doesn’t seem to be an issue on OSMC

**Comment by dagwieers** [comment]
@samnazarko It's weird as in OSMC it maps to KEY_TITLE, which seems better than KEY_EPG. However you added those eventmaps to LibreELEC: https://github.com/LibreELEC/LibreELEC.tv/blob/master/packages/sysutils/eventlircd/evmap/osmc_rf2.evmap
So I had high hopes you know why that difference :-) Maybe it is as simple as fixing LibreELEC to use KEY_TITLE instead? (I can't make it work with another **remote.xml** or fixes in **lircmap.xml**). And the file is on a read-only image so I can't influence it.

**Comment by samnazarko** [comment]
>And the file is on a read-only image so I can't influence it.

You should be able to get around this by using mount --bind and restarting eventlircd. This will let you test an updated keymap.

**Comment by dagwieers** [comment]
@samnazarko So I found why **remote.xml** failed to work, there was an XML-nesting typo. And with the existing eventmap it works if I use this as **remote.xml**:
```xml
<keymap>
  <global>
    <remote>
      <epg>ContextMenu</epg>
      <title>ContextMenu</title>
    </remote>
  </global>
</keymap>
```
And only `<epg>ContextMenu</epg>` is needed for the default eventmap, although I prefer to make the eventmap in LibreELEC identical to OSMC to be honest.

But it doesn't seem to work when the eventmap is mapping KEY_C to KEY_TITLE as KEY_TITLE doesn't seem to do anything by default. Which means we still have to add the **remote.xml** with `<title>ContextMenu</title>` to get it to work.

So I wonder how that would work in OSMC ?

**Comment by dagwieers** [comment]
What does work is mapping the OSMC remote in eventircd to KEY_CONTEXT_MENU directly.

This is unneeded for the first model (simply disable KEY_CONTEXT_MENU), but for the second model what is needed is `KEY_C = KEY_CONTEXT_MENU` and in this case not special mapping is needed in Kodi. So my issue when debugging this was the syntax error in **remote.xml**, and thanks to the bind-mount I was able to trace this to the source ;-)

The only remaining question is, how does OSMC work around this ? How is KEY_TITLE mapped to ContextMenu if there's no customized **remote.xml** ?

**Comment by DBMandrake** [comment]
@dagwieers 

Without looking, I suspect LibreElec don't use linux-input-layer in /usr/share/kodi/system/Lircmap.xml as we do.

In /lib/udev/rules.d/98-eventlircd.rules we have:

```
# change remote name sent by eventlircd to kodi from devinput to linux-input-layer
# for optimal button mapping with the default kodi Lircmap.xml
ENV{eventlircd_remote}="linux-input-layer"
```
This causes eventlircd to always pass linux-input-layer as the "name" of the remote device to Kodi. This causes Kodi to use the linux-input-layer section of Lircmap.xml for mapping all remote controls that are processed and handled by eventlircd.

This was a conscious decision by us right back in the beginning when we were first implementing remote support on OSMC. I suspect LibreElec just use the default of "devinput" which will have a reduced and different set of mappings.

**Comment by DBMandrake** [comment]
>That said, why wouldn't we want to have it expose KEY_CONTEXT_MENU directly ?
Or differently put, why does it map to KEY_EPG on LibreELEC and KEY_TITLE on OSMC (even if there's a layer correcting this) ?

The short answer is that KEY_CONTEXT_MENU is not listed in the default Lircmap.xml, therefore is not recognised by Kodi at all when sent over a Lirc socket from eventlircd. Only key bindings listed in Lircmap.xml can be used, and specifically those in the linux-input-layer section, which is the one we chose to use with eventlircd.

Today there is a KEY_MENU which maps to <menu> in Lircmap.xml and <menu> maps to ContextMenu in remote.xml, but from memory that mapping didn't exist 3 years ago in the default Kodi configuration files when we first designed the remote input system in OSMC.

So in the interests of not patching Kodi's out of the box remote.xml and Lircmap.xml we decided to define our remote to work correctly with the default config files so we're not constantly patching remote.xml and having to check it whenever there are changes to Kodi. So that meant defining the menu button on the remote to send KEY_TITLE.

Today it's only that way for legacy reasons. It could now be KEY_MENU and would work the same but it wouldn't offer any benefit to change it.

Lirc support is being removed from the latest Linux kernels so we will have to do a major redesign to the entire remote subsystem in OSMC soon so there is no point making small changes like changing KEY_TITLE to KEY_MENU now when there is a lot of upheaval coming soon to remote configurations.
